worker-message-prioritization-distribution/src/main/java/com/github/workerframework/workermessageprioritization/redistribution/lowlevel/StagingQueueTargetQueuePair.java

```diff
@@ -86,42 +146,58 @@ public void handleStagedMessage(final String consumerTag, final Envelope envelop
                     .priority(properties.getPriority())
                     .build();
 
-            //Hack the To See https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=614206
-            final JsonObject jsonObject =
-                    gson.fromJson(new String(body, StandardCharsets.UTF_8), JsonObject.class);
-            jsonObject.addProperty("to", targetQueue.getName());
-            final String s = gson.toJson(jsonObject);
-            //End Hack
+            targetQueueChannel.basicPublish("", targetQueue.getName(), basicProperties, body);
+        } catch (final IOException basicPublishException) {
 
-            targetQueueChannel.basicPublish("", targetQueue.getName(), basicProperties, 
-                    s.getBytes(StandardCharsets.UTF_8));
-        }
-        catch (final IOException e) {
-            LOGGER.error("Exception publishing to '{}' {}", targetQueue.getName(),
-                    e.toString());
-            stagingQueueChannel.basicCancel(consumerTag);
+            LOGGER.error("Exception publishing to '{}' for StagingQueueTargetQueuePair '{}' {}",
+                    targetQueue.getName(), this, basicPublishException.toString());
 
+            cancelConsumer(consumerTag);
         }
 
         messageCount.incrementAndGet();
 
-        if(messageCount.get() >= consumptionLimit) {
-            LOGGER.info("Consumption target '{}' reached for '{}'.", consumptionLimit,
-                    stagingQueue.getName());
-            if (stagingQueueConsumer.isCancelled()) {
-                stagingQueueChannel.basicCancel(consumerTag);
-            }
+        if (messageCount.get() >= consumptionLimit) {
+
+            LOGGER.debug("Consumption target '{}' reached for '{}'. Number of messages consumed by this consumer was '{}'. " +
+                            "Calling stagingQueueChannel.basicCancel({}) to cancel this consumer. " +
+                            "The StagingQueueTargetQueuePair this message relates to is '{}'",
+                    consumptionLimit,
+                    stagingQueue.getName(),
+                    messageCount.get(),
+                    consumerTag,
+                    this);
+
+            cancelConsumer(consumerTag);
         }
-        
     }
-    
+
+    private void cancelConsumer(final String consumerTag)
+    {
+        try {
+            stagingQueueConsumer.recordCancellationRequested();
+
+            stagingQueueChannel.basicCancel(consumerTag);
```

Comment:
Verify that any delivered messages, awaiting a basicAck are requeued immmediately when the channel is closed.

Reply:
Perhaps, requesting cancel should also be closing the channels?

Reply:
Talked about in previous comment: https://github.com/WorkerFramework/worker-message-prioritization/pull/41#discussion_r1246728921